### PR TITLE
refactor(Spanner): update samples to follow new samples guide

### LINKED ISSUE #1136: Spanner: samples should follow the samples guide
All the sample should follow the samples guide, this will also remove the command line argument.

### PR DESCRIPTION
**Pull request by laljikanjareeya**
This PR contains following Samples:

- spanner_add_timestamp_column
- spanner_create_table_with_datatypes
- spanner_create_table_with_timestamp_column
- spanner_insert_datatypes_data
- spanner_write_data_for_struct_queries
- spanner_query_with_array_parameter
- spanner_query_with_bool_parameter
- spanner_query_with_bytes_parameter
- spanner_query_with_date_parameter
- spanner_query_with_float_parameter
- spanner_query_with_int_parameter
- spanner_query_with_string_parameter
- spanner_query_with_timestamp_parameter
- spanner_retry_strategy

Fixes #1136

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
I'll get to this next week.

**Comment by amanda-tarafa** [comment]
Ping for myself for early next week.

**Comment by laljikanjareeya** [comment]
@amanda-tarafa Ping for review.

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs:62]: nit: Indents seem too big

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs:64]: nit: having several of these in on line is not common, maybe one per line.

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Mostly nits.
Although there are some changes in the fixture that will clash with #1315. FYI @olavloite and @skuruppu.

Also some of the samples appear not to have tests, only the query samples have tests I think.

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:337]: `Venues` Table has been created in fixture with the use of sample as it is part of [CreateTableWithDatatypesAsync](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/1304/files#diff-889532a8f286d9a39a3ef4e4f07bf6733bb716cd46a3ea495a9e379e32dbd4bf) sample.

**Comment by olavloite** [comment]
> Although there are some changes in the fixture that will clash with #1315. FYI @olavloite and @skuruppu.
> 

Thanks for the notification. As far as I'm concerned, you don't have to let any potential clashes with #1315 be a blocker from merging this PR. I can update #1315 at a later moment if needed, and it might take some time before we can merge #1315.

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/RetryRobot.cs:20]: @skuruppu Do we need this Sample?

**Comment by amanda-tarafa** [comment]
> > Although there are some changes in the fixture that will clash with #1315. FYI @olavloite and @skuruppu.
> 
> Thanks for the notification. As far as I'm concerned, you don't have to let any potential clashes with #1315 be a blocker from merging this PR. I can update #1315 at a later moment if needed, and it might take some time before we can merge #1315.

OK, then we can merge this one when it's ready, thanks both!

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/RetryRobot.cs]: Done!

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/RetryRobot.cs]: Done!

File: spanner/api/Spanner.Samples/QueryWithIntAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/QueryWithIntAsync.cs]: Done!

File: spanner/api/Spanner.Samples/QueryWithIntAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/QueryWithIntAsync.cs]: Done!

File: spanner/api/Spanner.Samples/QueryWithFloatAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/QueryWithFloatAsync.cs]: Done!

File: spanner/api/Spanner.Samples.Tests/QueryWithFloatAsyncTest.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples.Tests/QueryWithFloatAsyncTest.cs]: Done!

File: spanner/api/Spanner.Samples.Tests/QueryWithIntAsyncTest.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples.Tests/QueryWithIntAsyncTest.cs]: Done and Check in other Test.

File: spanner/api/Spanner.Samples/QueryWithArrayAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/QueryWithArrayAsync.cs]: Done, Also check in other sample. 👍

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs:65]: Done!

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs]: Done!

**Comment by laljikanjareeya** [comment]
> Also some of the samples appear not to have tests, only the query samples have tests I think.

Some of the samples don't have separate tests as it's included in the fixture if it fails all the sample's tests will falil.
writing tests for these samples might increase the test time because It's needed to create a separate database for each test.

File: spanner/api/Spanner.Samples.Tests/QueryDataWithNumericParameterTest.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples.Tests/QueryDataWithNumericParameterTest.cs:35]: waiting for the conclusion.

Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/QueryDataWithNumericParameterTest.cs:35]: #1315 will be updated to include these changes so we can leave as is.

File: spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs:19]: Ping on this. Not sure why we replaced the numeric test with this other. Don't we need both?

File: spanner/api/Spanner.Samples.Tests/SpannerFixture.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/SpannerFixture.cs:337]: #1315 will be updated to include these changes so we can leave this one as is.

File: spanner/api/Spanner.Samples.Tests/UpdateDataWithNumericTest.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/UpdateDataWithNumericTest.cs:31]: #1315 Will be updated to include these changes so we can leave this one as is.

File: spanner/api/Spanner.Samples.Tests/WriteDataWithTimestampAsyncTest.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples.Tests/WriteDataWithTimestampAsyncTest.cs:32]: #1315 Will be updated to include these changes so we can leave this one as is.

File: spanner/api/Spanner.Samples/UpdateDataWithNumericAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/UpdateDataWithNumericAsync.cs:46]: nit: all these indent changes seem wrong

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs:24]: nit: Datatypes => DataTypes, here and everywhere.

File: spanner/api/Spanner.Samples/QueryWithTimestampAsync.cs
Comment (amanda-tarafa) [inline spanner/api/Spanner.Samples/QueryWithTimestampAsync.cs:35]:
```suggestion
        DateTime exampleTimestamp = DateTime.UtcNow;
```

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/RetryRobot.cs:20]: @skuruppu ping on this one, thanks!

**Review by amanda-tarafa — COMMENTED** [review]
A couple of nits but otherwise LGTM.
I'll still wait for @skuruppu to reply to the open questions and review before doing a final review of possible changes.

File: spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs
Comment (skuruppu) [reply spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs:19]: Agree, I think we need both.

File: spanner/api/Spanner.Samples/CreateTableWithDatatypesAsync.cs
Comment (skuruppu) [inline spanner/api/Spanner.Samples/CreateTableWithDatatypesAsync.cs:37]: There should be a `Revenue NUMERIC` column here too. Is this coming in a later PR?

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (skuruppu) [reply spanner/api/Spanner.Samples/RetryRobot.cs:20]: I'm afraid that there's a whole section on the [Getting Started guide](https://cloud.google.com/spanner/docs/getting-started/csharp#performing_retries_on_read-write_transactions) which specifically talk about retries in C#. So we'll need to keep this.

**Review by skuruppu — CHANGES_REQUESTED** [review]

File: spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples.Tests/QueryWithBoolAsyncTest.cs:19]: Kept NumericColumn Test as it is.

File: spanner/api/Spanner.Samples/UpdateDataWithNumericAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/UpdateDataWithNumericAsync.cs]: Done!

File: spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/InsertDatatypesDataAsync.cs]: Done and check everywhere.

File: spanner/api/Spanner.Samples/QueryWithTimestampAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/QueryWithTimestampAsync.cs]: Done!

File: spanner/api/Spanner.Samples/CreateTableWithDatatypesAsync.cs
Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/CreateTableWithDatatypesAsync.cs:37]: Added `Revenue` Column.

**Review by skuruppu — APPROVED** [review]

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/RetryRobot.cs:20]: Can we at least add the Spanner specific exceptions that are retryable. As far as I can see this is just a generic retry code.

Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/RetryRobot.cs:21]: Done!

File: spanner/api/Spanner.Samples/QueryWithIntAsync.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/QueryWithIntAsync.cs]: @skuruppu ping on this?

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (amanda-tarafa) [reply spanner/api/Spanner.Samples/RetryRobot.cs:21]: @skuruppu Looking at the docs that you linked above, I really think we should be showing a retriable transaction example instead of this retry robot. It might be that the retry robot was implemented before we had the retriable transaction, but it seems outdated now. WDYT?

**Review by amanda-tarafa — COMMENTED** [review]
Code LGTM, just a couple questions pending. Thanks

File: spanner/api/Spanner.Samples/RetryRobot.cs
Comment (skuruppu) [reply spanner/api/Spanner.Samples/RetryRobot.cs:21]: Hmmm this is weird in the sense that such a retry example doesn't exist for other languages. It's possible it's because they all retry aborted transactions internally.

I don't think we need this example at all. Instead, what I'll do is to modify one of the other samples on that page to use a retriable transaction and modify the text to say something like "BTW, this is how you can retry Aborted transactions".

So @laljikanjareeya please delete this file from this PR.

Comment (laljikanjareeya) [reply spanner/api/Spanner.Samples/RetryRobot.cs]: Removed!

**Review by amanda-tarafa — APPROVED** [review]
LGTM.

Will wait for tests to go green before merging.

**Comment by amanda-tarafa** [comment]
There's a Spanner flake but on a test not related to this PR, I've created #1332.

The other failure is unrelated and being tracked in #1311.

I'll force merge.

**Pull request merged**
